enarx-keep-sev/src/vm.rs

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::x86_64::*;
+
+use std::error::Error;
+
+use kvm_bindings::kvm_userspace_memory_region;
+use kvm_ioctls::{Kvm, VmFd};
+use memmap::*;
+
+pub struct VirtualMachine {
+    kvm: Kvm,
+    kvm_fd: VmFd,
+    address_space: Vec<kvm_userspace_memory_region>,
+    mmap_handles: Vec<MmapMut>,
+}
+
+impl VirtualMachine {
+    pub fn new(mem_size: usize) -> Result<Self, Box<dyn Error>> {
+        // Create a KVM context
+        let kvm = Kvm::new()?;
+        let kvm_fd = kvm.create_vm()?;
+
+        // Create the guest address space
+        let guest_addr_space = memmap::MmapOptions::new().len(mem_size).map_anon()?;
+        let region = kvm_userspace_memory_region {
+            slot: 0,
+            flags: 0,
+            guest_phys_addr: 0,
+            memory_size: mem_size as _,
+            userspace_addr: &*guest_addr_space as *const _ as *const u8 as u64,
+        };
+
+        unsafe {
+            kvm_fd.set_user_memory_region(region)?;
+        }
+
+        // Set up the page tables
+        let mut page_tables = PageTables::default();
+        page_tables.pml4t[0] = PDPTE_START as u64 | 0x7;
+        page_tables.pml3t_ident[0] = PDE_START as u64 | 0x7;
+        page_tables.pml2t_ident[0] = 0x183u64;
```

Comment:
```suggestion
        // Set up the page tables
        page_tables.pml4t[0] = PDPTE_START as u64 | 0x3; // page flags WRITABLE | PRESENT
        page_tables.pml3t_ident[0] = 0x83u64; // page flags HUGE_PAGE | WRITABLE | PRESENT
```

The shim doesn't need it that fine grained anymore.

You may also want to substitute the flags with some bitfield values.

Reply:
https://github.com/rust-osdev/x86_64/blob/master/src/structures/paging/page_table.rs#L107-L169
Now, that I patched the `x86_64` crate to be usable with non-nightly rust.
```toml
x86_64 = { version = "0.11.0", default-features = false, features = ["stable"] }
```

Reply:
https://docs.rs/x86_64/0.11.0/x86_64/structures/paging/page_table/struct.PageTableFlags.html